docs/stdout_publisher.md

```diff
@@ -1,50 +1,73 @@
-# Example usage with [telegraf]
+# Using the stdout publisher
 
-You can use this project in combination with [telegraf] or similar tools to get
-your inverter's data into other systems than mqtt.
+Enabling this with `DEYE_FEATURE_STDOUT_PUBLISHER` will dump all collected
+metrics in json on stdout once every cycle (every `DEYE_DATA_READ_INTERVAL`
+seconds).
+This can be useful in combination with other tools to process and push the
+metrics into different systems other than mqtt.
 
-An example usage with [telegraf] could look something like that:
+The json object has the following structure:
+```json
+{
+  "serial": "1234567890",
+  "address": "192.168.0.1",
+  "port": 8899,
+  "data": [
+    {
+      "up": 1,
+    }, {
+      "temp": 31,
```

Comment:
The value property name is not consistent across metrics. Here it's `temp`, next it's `energy`. Wouldn't it be simpler to use constant property name, e.g. `value`?

Reply:
This plays very nice with using telegraf to collect and process those metrics.

telegraf, with the [json_v2 parser](https://github.com/influxdata/telegraf/tree/master/plugins/parsers/json_v2), will consider (pretty much) everything that is not used as [a `timestamp`](https://github.com/kbialek/deye-inverter-mqtt/pull/90/files#diff-008cb0b3f78bb87f24168135b73bc2aba1a25d00eafb89467386ec1d2d942755R85) or as [a `tag`](https://github.com/kbialek/deye-inverter-mqtt/pull/90/files#diff-008cb0b3f78bb87f24168135b73bc2aba1a25d00eafb89467386ec1d2d942755R87) as a `field` (= "metric value").

Essentially the `field` in combination with [the `measurement_name`](https://github.com/kbialek/deye-inverter-mqtt/pull/90/files#diff-008cb0b3f78bb87f24168135b73bc2aba1a25d00eafb89467386ec1d2d942755R76) builds up the metric name. So that leaves me with the following metrics / time-series when pushed to prometheus:
```
deye_up{serial="123123"} 1
deye_energy{serial="123123", source="dc/pv1", unit="W", ...} 123
deye_energy{serial="123123", source="dc/pv2", unit="W", ...} 99
deye_current{serial="123123", source="dc/pv1", unit="A", ...} 2.1
deye_current{serial="123123", source="dc/pv2", unit="A", ...} 1.2
...
```
So one time series per "dimension" (energy, up, current, ...), with different tags (telegraf speak) = labels (prometheus speak).

If we'd use the a generic `value` key in the JSON object, we'd probably also need a `name`/`metric_name`/... key. Then, for my use-case, I'd probably need a processor step, which transforms from e.g. `deye_value{serial="123123", metric_name="energy", ...} 0.3` to `deye_energy{serial="123123", ...} 0.3`.

Now, I am not generally against using a generic `value` field.

I wouldn't do it, however, because
- this works nice with the use-case at hand
- I don't know of other use-cases
- I would rather "optimize" for the use-case I know of
    - rather than for a hypothetical, yet unknown, use-case
    - having said that, of course this should not block hypothetical future use-cases either
        - which it doesn't ; other consumers, if that data-model does not fit them, can still transform the data as needed


Again, not against it, just laying out my thoughts and reasoning.

Reply:
Okay, then perhaps this publisher should be renamed to `TelegrafJsonV2Publisher`. WDYT?

Reply:
I am not sure, that's a mouth full, TBH, and too specific to telegraf. The output can still be consumed by anything that can parse json, not just telegraf. And that's definitely one of my goals. It "just happens" to work really nice with telegraf/json_v2 out of the box.

What do you think about
- keeping the name as is
- explicitly call out in the docs that
    - this can be used with anything that can parse json
    - however, it uses a format that works really well with telegraf/json_v2
    - if that format does not work for specific use-cases, ask users to open issues

If we find real good evidence that another format is needed, we can implement another config knob (e.g. `STDOUT_FORMAT`) in a backwards compatible way. I am not saying we should, just that we could, if the need should arise.

Reply:
Let me ask differently. Who is backing this format? Is it Influx? https://docs.influxdata.com/telegraf/v1.21/data_formats/input/json_v2/

Chosen name should be specific and self explanatory.

> If we find real good evidence that another format is needed, we can implement another config knob (e.g. STDOUT_FORMAT) in a backwards compatible way. I am not saying we should, just that we could, if the need should arise.

This is an illusive future requirement. Don't take it personally. Imho, devs have a tendency to anticipate such imagined features and as a result drag unnecessary complexity into the projects. Therefore, if such a requirement materialized, my approach would be to implement a new publisher, and provide a migration path. And as for now, I strongly prefer to keep it possibly simple.

Reply:
> Let me ask differently. Who is backing this format? Is it Influx?

I guess a combination of the implementation of json_v2 in telegraf and [gjson](https://github.com/tidwall/gjson/). However, it's able to consume various different layouts of json documents. Further down you can see an example with a different input format and a changed telegraf config which would still yield the same data.

The point I am trying to make is: It's just json. Yes, it plays nice with telegraf/json_v2, but it is still telegraf would be capable of consuming different layouts of the json document on the one hand, and any other tool, say `jq` for example, would be able to consume and make sense of the currently proposed layout too. Thus I am not convinced to call it `TelegrafJsonV2Publisher` as neither telegraf nor json_v2 really define any specifics, but can be configured to work with various different layouts. Therefore I argue this is a 'generic layout' that any tool that is capable of parsing json should be able to consume and make sense of.

Having said that, in the end I don't care how it's called. You make the call.

But coming back to the "main point":
Would you rather prefer objects like:
```json
{
  "value": 23,   # <- the key is always the generic "value"
  "metric": "temp",  # <- the actual "metric name" is in another field
  "unit": "something",
    ...
}
```
or
```json
{
  "temp": 23,   # <-- the key is the "metric name"
  "unit": "something",
    ...
}
```
I'd prefer the later, which is what the current implementation implements. But I think you lean towards the former, is that right?

> This is an illusive future requirement. Don't take it personally. Imho, devs have a tendency to anticipate such imagined features and as a result drag unnecessary complexity into the projects. Therefore, if such a requirement materialized, my approach would be to implement a new publisher, and provide a migration path. And as for now, I strongly prefer to keep it possibly simple.

I think we are on the same page here, generally.
- I don't see a need to implement any other output format, the current one should be able to handle anything if the user's tool is capable of parsing json. Any transformation into a destination system's specifics needs to be handled by the consumer anyway.
- If I am wrong, and there is a need for a different layout / format this can still be implemented. Nothing that this PR introduces would block that; that's pretty much the only point I tried to make
- If a hypothetical future output format would be implemented in `DeyeStdoutPublisher` or it's own thing, how it would be configured, ... that's all implementation details IMHO.

<details>
<summary>Example with different json object layout</summary>

If `DeyeStdoutPublisher` would produce slightly different output like
```json
{
  "serial": "1234567890",
  "address": "192.168.0.1",
  "port": 8899,
  "data": [
    {
      "value": 1,
      "metric": "up"
    }, {
      "value": 31,
      "metric": "temp",
      "name": "Radiator temperature",
      "unit": "°C",
      "groups": "micro",
      "sensor_type": "SingleRegisterSensor",
      "sensor": "radiator",
      "timestamp": 1686219130
    }, {
      "value": 0.2,
      "metric": "energy",
      "name": "PV1 Total",
      "unit": "kWh",
      "groups": "micro",
      "sensor_type": "DoubleRegisterSensor",
      "sensor": "dc/pv1/total",
      "timestamp": 1686219130
    }, {
      ...
    }
  ]
}
```
telegraf with json_v2 would still be able to consume this with e.g. a config something like that:
```toml
[[inputs.file]]
  files = ["new.json"]
  data_format = "json_v2"

  [[inputs.file.json_v2]]
    measurement_name = "deye"

    [[inputs.file.json_v2.object]]
      path = "data"

      timestamp_key = "timestamp"
      timestamp_format = "unix"

      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="power")#.value'
        rename = "power"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="energy")#.value'
        rename = "energy"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="up")#.value'
        rename = "up"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="temp")#.value'
        rename = "temp"

      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor'
      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor_type'
      [[inputs.file.json_v2.object.tag]]
        path = '#.unit'
      [[inputs.file.json_v2.object.tag]]
        path = '#.groups'
      [[inputs.file.json_v2.object.tag]]
        path = '#.name'

    [[inputs.file.json_v2.tag]]
      path = 'address'
    [[inputs.file.json_v2.tag]]
      path = 'port'
    [[inputs.file.json_v2.tag]]
      path = 'serial'
```

Note: I assume, this can  be made way nicer, if one understands how gjson really works, which I don't.

</details>

Reply:
> Let me ask differently. Who is backing this format? Is it Influx?

I guess a combination of the implementation of json_v2 in telegraf and [gjson](https://github.com/tidwall/gjson/). However, it's able to consume various different layouts of json documents. Further down you can see an example with a different input format and a changed telegraf config which would still yield the same data.

The point I am trying to make is: It's just json. Yes, it plays nice with telegraf/json_v2, but it is still telegraf would be capable of consuming different layouts of the json document on the one hand, and any other tool, say `jq` for example, would be able to consume and make sense of the currently proposed layout too. Thus I am not convinced to call it `TelegrafJsonV2Publisher` as neither telegraf nor json_v2 really define any specifics, but can be configured to work with various different layouts. Therefore I argue this is a 'generic layout' that any tool that is capable of parsing json should be able to consume and make sense of.

Having said that, in the end I don't care how it's called. You make the call.

But coming back to the "main point":
Would you rather prefer objects like:
```json
{
  "value": 23,   # <- the key is always the generic "value"
  "metric": "temp",  # <- the actual "metric name" is in another field
  "unit": "something",
    ...
}
```
or
```json
{
  "temp": 23,   # <-- the key is the "metric name"
  "unit": "something",
    ...
}
```
I'd prefer the later, which is what the current implementation implements. But I think you lean towards the former, is that right?

> This is an illusive future requirement. Don't take it personally. Imho, devs have a tendency to anticipate such imagined features and as a result drag unnecessary complexity into the projects. Therefore, if such a requirement materialized, my approach would be to implement a new publisher, and provide a migration path. And as for now, I strongly prefer to keep it possibly simple.

I think we are on the same page here, generally.
- I don't see a need to implement any other output format, the current one should be able to handle anything if the user's tool is capable of parsing json. Any transformation into a destination system's specifics needs to be handled by the consumer anyway.
- If I am wrong, and there is a need for a different layout / format this can still be implemented. Nothing that this PR introduces would block that; that's pretty much the only point I tried to make
- If a hypothetical future output format would be implemented in `DeyeStdoutPublisher` or it's own thing, how it would be configured, ... that's all implementation details IMHO.

<details>
<summary>Example with different json object layout</summary>

If `DeyeStdoutPublisher` would produce slightly different output like
```json
{
  "serial": "1234567890",
  "address": "192.168.0.1",
  "port": 8899,
  "data": [
    {
      "value": 1,
      "metric": "up"
    }, {
      "value": 31,
      "metric": "temp",
      "name": "Radiator temperature",
      "unit": "°C",
      "groups": "micro",
      "sensor_type": "SingleRegisterSensor",
      "sensor": "radiator",
      "timestamp": 1686219130
    }, {
      "value": 0.2,
      "metric": "energy",
      "name": "PV1 Total",
      "unit": "kWh",
      "groups": "micro",
      "sensor_type": "DoubleRegisterSensor",
      "sensor": "dc/pv1/total",
      "timestamp": 1686219130
    }, {
      ...
    }
  ]
}
```
telegraf with json_v2 would still be able to consume this with e.g. a config something like that:
```toml
[[inputs.file]]
  files = ["new.json"]
  data_format = "json_v2"

  [[inputs.file.json_v2]]
    measurement_name = "deye"

    [[inputs.file.json_v2.object]]
      path = "data"

      timestamp_key = "timestamp"
      timestamp_format = "unix"

      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="power")#.value'
        rename = "power"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="energy")#.value'
        rename = "energy"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="up")#.value'
        rename = "up"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="temp")#.value'
        rename = "temp"

      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor'
      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor_type'
      [[inputs.file.json_v2.object.tag]]
        path = '#.unit'
      [[inputs.file.json_v2.object.tag]]
        path = '#.groups'
      [[inputs.file.json_v2.object.tag]]
        path = '#.name'

    [[inputs.file.json_v2.tag]]
      path = 'address'
    [[inputs.file.json_v2.tag]]
      path = 'port'
    [[inputs.file.json_v2.tag]]
      path = 'serial'
```

Note: I assume, this can  be made way nicer, if one understands how gjson really works, which I don't.

</details>

Reply:
> Let me ask differently. Who is backing this format? Is it Influx?

I guess a combination of the implementation of json_v2 in telegraf and [gjson](https://github.com/tidwall/gjson/). However, it's able to consume various different layouts of json documents. Further down you can see an example with a different input format and a changed telegraf config which would still yield the same data.

The point I am trying to make is: It's just json. Yes, it plays nice with telegraf/json_v2, but it is still telegraf would be capable of consuming different layouts of the json document on the one hand, and any other tool, say `jq` for example, would be able to consume and make sense of the currently proposed layout too. Thus I am not convinced to call it `TelegrafJsonV2Publisher` as neither telegraf nor json_v2 really define any specifics, but can be configured to work with various different layouts. Therefore I argue this is a 'generic layout' that any tool that is capable of parsing json should be able to consume and make sense of.

Having said that, in the end I don't care how it's called. You make the call.

But coming back to the "main point":
Would you rather prefer objects like:
```json
{
  "value": 23,   # <- the key is always the generic "value"
  "metric": "temp",  # <- the actual "metric name" is in another field
  "unit": "something",
    ...
}
```
or
```json
{
  "temp": 23,   # <-- the key is the "metric name"
  "unit": "something",
    ...
}
```
I'd prefer the later, which is what the current implementation implements. But I think you lean towards the former, is that right?

> This is an illusive future requirement. Don't take it personally. Imho, devs have a tendency to anticipate such imagined features and as a result drag unnecessary complexity into the projects. Therefore, if such a requirement materialized, my approach would be to implement a new publisher, and provide a migration path. And as for now, I strongly prefer to keep it possibly simple.

I think we are on the same page here, generally.
- I don't see a need to implement any other output format, the current one should be able to handle anything if the user's tool is capable of parsing json. Any transformation into a destination system's specifics needs to be handled by the consumer anyway.
- If I am wrong, and there is a need for a different layout / format this can still be implemented. Nothing that this PR introduces would block that; that's pretty much the only point I tried to make
- If a hypothetical future output format would be implemented in `DeyeStdoutPublisher` or it's own thing, how it would be configured, ... that's all implementation details IMHO.

<details>
<summary>Example with different json object layout</summary>

If `DeyeStdoutPublisher` would produce slightly different output like
```json
{
  "serial": "1234567890",
  "address": "192.168.0.1",
  "port": 8899,
  "data": [
    {
      "value": 1,
      "metric": "up"
    }, {
      "value": 31,
      "metric": "temp",
      "name": "Radiator temperature",
      "unit": "°C",
      "groups": "micro",
      "sensor_type": "SingleRegisterSensor",
      "sensor": "radiator",
      "timestamp": 1686219130
    }, {
      "value": 0.2,
      "metric": "energy",
      "name": "PV1 Total",
      "unit": "kWh",
      "groups": "micro",
      "sensor_type": "DoubleRegisterSensor",
      "sensor": "dc/pv1/total",
      "timestamp": 1686219130
    }, {
      ...
    }
  ]
}
```
telegraf with json_v2 would still be able to consume this with e.g. a config something like that:
```toml
[[inputs.file]]
  files = ["new.json"]
  data_format = "json_v2"

  [[inputs.file.json_v2]]
    measurement_name = "deye"

    [[inputs.file.json_v2.object]]
      path = "data"

      timestamp_key = "timestamp"
      timestamp_format = "unix"

      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="power")#.value'
        rename = "power"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="energy")#.value'
        rename = "energy"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="up")#.value'
        rename = "up"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="temp")#.value'
        rename = "temp"

      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor'
      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor_type'
      [[inputs.file.json_v2.object.tag]]
        path = '#.unit'
      [[inputs.file.json_v2.object.tag]]
        path = '#.groups'
      [[inputs.file.json_v2.object.tag]]
        path = '#.name'

    [[inputs.file.json_v2.tag]]
      path = 'address'
    [[inputs.file.json_v2.tag]]
      path = 'port'
    [[inputs.file.json_v2.tag]]
      path = 'serial'
```

Note: I assume, this can  be made way nicer, if one understands how gjson really works, which I don't.

</details>

Reply:
Thank you for your elaborated answers. After reading that I came to the following conclusion:
1. `json_v2` is not a data format, but a parser spec used by the data consumer. So it does not matter much how the output JSON will look like, as the client can still be configured to digest it. Hence, further discussions about the JSON format are pointless.
2. The functionality that you've implemented does not fit into the scope of this project. IIRC, I've already told that before when you came with the idea, and now I'm very confident about that. I prefer to follow the Unix philosophy of doing one thing and doing it well. On the other hand this same philosophy also promotes extensibility.
3. So, I'd like to help you to achieve your goal. I've decided to add a simple plugin system that will allow you to do what you want, as you like, without asking anyone for approval. I encourage you to give it a try and transform your code into the plugin. See https://github.com/kbialek/deye-inverter-mqtt/tree/feature/plugin-loader IMO, this change will let us keep the things maintainable and properly decoupled. 
4. In case you want to make your work public, then start a github project. I will be happy to link it in the readme.

Reply:
Ok then.
Thanks for the plugin mechanism!

Reply:
> Thanks for the plugin mechanism!

No problem. Please let me know if it works for you as expected.

Reply:
Yeah, I am running with `ghcr.io/hoegaarden/deye-inverter-mqtt-plugins:55db5c1` from [here](https://github.com/hoegaarden/deye-inverter-mqtt-plugins/) -- works for me.